README.rst
Django Scaler
=============
**Degrade gracefully by automatically replacing heavy pages with static pages while a server is taking strain.**

.. contents:: Contents
    :depth: 5

Installation
------------

#. Install or add ``django-scaler`` to your Python path.

#. Add ``scaler`` to your ``INSTALLED_APPS`` setting.

#. Add ``scaler.middleware.ScalerMiddleware`` to the top of your ``MIDDLEWARE_CLASSES`` setting.

#. Add (r'^scaler/', include('scaler.urls')) to urlpatterns.

Overview
--------

Servers may at times get overloaded due to a variety of reasons. When that
happens you don't want expensive requests to bring down your entire site. The
site must redirect expensive requests to a "server busy" page while the server
is under load, and then automatically remove the redirects once the load has
dropped enough. 

`django-scaler` addresses this situation in two ways. Firstly, it knows which
requests to redirect by itself. Secondly, it can be instructed to redirect the
N most expensive requests. It stores response time data in in-memory caches
enabling it to make these decisions.

Usage
-----

Pasted from ``test_settings.py``::

    DJANGO_SCALER = { 
        'server_busy_url_name': 'server-busy',

        # How many response times to consider for an URL. A small value means slow
        # response times are quickly acted upon, but it may be overly aggressive. 
        # A large value means an URL must be slow for a number of requests before 
        # it is acted upon. The default is 100.
        'trend_size': 10,

        # How much slower than average the trend must be before redirection kicks
        # in. The default is 4.0.
        'slow_threshold': 2.0,

        # How many seconds to keep redirecting an URL before serving normally. The
        # default is 60.
        'redirect_for': 10,

        # A function that returns how many of the slowest URLs must be redirected.
        # Depending on the site, data and load on the server this may be a large
        # number. This allows external processes to instruct the middleware to
        # redirect. The default is 0.
        'redirect_n_slowest_function': lambda: 0,

        # A function that returns what percentage of the slowest URLs must be
        # redirected. Depending on the site, data and load on the server this may
        # approach 100. The default is 0.
        'redirect_percentage_slowest_function': lambda: 0,

        # A function that returns a list of regexes. URLs matching the regexes are
        # redirected. Each regex is a simple string. Do not prefix with r''. The
        # default is an empty list.
        'redirect_regexes_function': lambda: [],
    }



CHANGELOG.rst
Changelog
=========

0.2
---
#. Regex URL matching for explicit redirection.
#. Percentage slowest explicit redirection.

0.1
---
#. First release.



scaler/tests.py
import time

from django.test import TestCase
from django.test.client import Client
from django.core.files import File as DjangoFile
from django.core.urlresolvers import reverse
from django.conf import settings


class ScalerTestCase(TestCase):

    def setUp(self):
        self.client = Client()

    def reset_settings(self):
        settings.DJANGO_SCALER['redirect_n_slowest_function'] = lambda: 0
        settings.DJANGO_SCALER['redirect_percentage_slowest_function'] = \
            lambda: 0
        settings.DJANGO_SCALER['redirect_regexes_function'] = lambda: []

    def test_auto_scaler(self):
        """Middleware redirects requests by itself"""
        self.reset_settings()

        # Use delay to smooth out anomalies. A page may render in 5ms and the
        # next time in 20ms because of many reasons. A large enough delay
        # flattens the discrepancy ratio.
        for i in range(0, 20):
            response = self.client.get('/?delay=0.1')
            self.assertEqual(response.status_code, 200)

        # Start making slow requests. We expect a redirection after X requests.
        # We don't care exactly when since many things may influence response
        # time. The math does indicate that redirection is expected sometime.
        for i in range(0, 10):
            response = self.client.get('/?delay=1.0')
            if response.status_code == 302:
                stamp = time.time()
        self.assertEqual(response.status_code, 302)

        # Make requests every 2 seconds. Redirects are expected for
        # 10 seconds.
        now = time.time()
        while now - settings.DJANGO_SCALER['redirect_for'] < stamp:
            response = self.client.get('/')
            self.assertEqual(response.status_code, 302)
            time.sleep(2)
            now = time.time()

        # Time has elapsed. Redirection is disabled for the next request.
        response = self.client.get('/')
        self.assertEqual(response.status_code, 200)

        # And the next request will pass as well since the trend cache is
        # clean.
        response = self.client.get('/')
        self.assertEqual(response.status_code, 200)

    def test_excplicit_scaler_n_slowest(self):
        """Middleware is instructed to redirect X slowest URLs"""
        self.reset_settings()

        # Do calls so we can decide which URLs are slowest.
        for i in range(0, 20):
            response = self.client.get('/?delay=0.1')
            self.assertEqual(response.status_code, 200)
            response = self.client.get('/scaler-test-one/?delay=0.3')
            self.assertEqual(response.status_code, 200)
            response = self.client.get('/scaler-test-two/?delay=0.5')
            self.assertEqual(response.status_code, 200)

        # Set the redirect_n_slowest_function
        settings.DJANGO_SCALER['redirect_n_slowest_function'] = lambda: 2

        response = self.client.get('/')
        self.assertEqual(response.status_code, 200)
        response = self.client.get('/scaler-test-one/')
        self.assertEqual(response.status_code, 302)
        response = self.client.get('/scaler-test-two/')
        self.assertEqual(response.status_code, 302)

    def test_excplicit_scaler_percentage_slowest(self):
        """Middleware is instructed to redirect % slowest URLs"""
        self.reset_settings()

        # Do calls so we can decide which URLs are slowest.
        for i in range(0, 20):
            response = self.client.get('/?delay=0.1')
            self.assertEqual(response.status_code, 200)
            response = self.client.get('/scaler-test-one/?delay=0.3')
            self.assertEqual(response.status_code, 200)
            response = self.client.get('/scaler-test-two/?delay=0.5')
            self.assertEqual(response.status_code, 200)

        # Set the redirect_percentage_slowest_function
        settings.DJANGO_SCALER['redirect_percentage_slowest_function'] = \
            lambda: 67
        response = self.client.get('/')
        self.assertEqual(response.status_code, 200)
        response = self.client.get('/scaler-test-one/')
        self.assertEqual(response.status_code, 302)
        response = self.client.get('/scaler-test-two/')
        self.assertEqual(response.status_code, 302)

    def test_excplicit_scaler_regexes(self):
        """Middleware is instructed to redirect if URL matches a regex in
        list."""
        self.reset_settings()

        settings.DJANGO_SCALER['redirect_regexes_function'] = \
            lambda: ['/scaler-test-o', '/scaler-test-t']
        response = self.client.get('/')
        self.assertEqual(response.status_code, 200)
        response = self.client.get('/scaler-test-one/')
        self.assertEqual(response.status_code, 302)
        response = self.client.get('/scaler-test-two/')
        self.assertEqual(response.status_code, 302)


test_settings.py
import os

DEBUG = True
TEMPLATE_DEBUG = DEBUG

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
    }
}

ROOT_URLCONF = 'scaler.urls'

INSTALLED_APPS = (
    'scaler',
    'django.contrib.auth',
    'django.contrib.admin',
    'django.contrib.contenttypes',
)

MIDDLEWARE_CLASSES = (
    'scaler.middleware.ScalerMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
)

TEMPLATE_CONTEXT_PROCESSORS = (
    'django.core.context_processors.debug',
    'django.core.context_processors.i18n',
    'django.core.context_processors.request',
)

DJANGO_SCALER = { 
    'server_busy_url_name': 'server-busy',

    # How many response times to consider for an URL. A small value means slow
    # response times are quickly acted upon, but it may be overly aggressive. 
    # A large value means an URL must be slow for a number of requests before 
    # it is acted upon. The default is 100.
    'trend_size': 10,

    # How much slower than average the trend must be before redirection kicks
    # in. The default is 4.0.
    'slow_threshold': 2.0,

    # How many seconds to keep redirecting an URL before serving normally. The
    # default is 60.
    'redirect_for': 10,

    # A function that returns how many of the slowest URLs must be redirected.
    # Depending on the site, data and load on the server this may be a large
    # number. This allows external processes to instruct the middleware to
    # redirect. The default is 0.
    'redirect_n_slowest_function': lambda: 0,

    # A function that returns what percentage of the slowest URLs must be
    # redirected. Depending on the site, data and load on the server this may
    # approach 100. The default is 0.
    'redirect_percentage_slowest_function': lambda: 0,

    # A function that returns a list of regexes. URLs matching the regexes are
    # redirected. Each regex is a simple string. Do not prefix with r''. The
    # default is an empty list.
    'redirect_regexes_function': lambda: [],
}


scaler/urls.py
from django.conf.urls.defaults import patterns, url


urlpatterns = patterns('',
    url(
        r'^$',
        'django.views.generic.simple.direct_to_template',
        {
            'template': 'scaler/test.html',
        },
        name='scaler-test'
    ),
    url(
        r'^scaler-test-one/$',
        'django.views.generic.simple.direct_to_template',
        {
            'template': 'scaler/test.html',
        },
        name='scaler-test-one'
    ),
    url(
        r'^scaler-test-two/$',
        'django.views.generic.simple.direct_to_template',
        {
            'template': 'scaler/test.html',
        },
        name='scaler-test-two'
    ),
    url(
        r'^server-busy/$',
        'django.views.generic.simple.direct_to_template',
        {
            'template': 'scaler/server_busy.html',
        },
        name='server-busy'
    ),
)


scaler/middleware.py
import time
import re

from django.http import HttpResponseRedirect
from django.core.cache import cache
from django.core.urlresolvers import reverse
from django.conf import settings

# In-memory caches are used since different processes do not necessarily
# exhibit the same response times, even though they may share a caching backend
# like memcached. We also don't have to be concerned with thread safety so no
# need to use LocMemCache.
_cache = {}
_request_response_times = {}

SERVER_BUSY_URL = reverse(
    settings.DJANGO_SCALER.get('server_busy_url_name', 'server-busy')
)


def redirect_n_slowest_dummy():
    return 0


def redirect_n_slowest_from_cache():
    """Simple retrieval from whatever cache is in use"""
    return cache.get('django_scaler_n_slowest')


def redirect_percentage_slowest_dummy():
    return 0


def redirect_percentage_slowest_from_cache():
    return cache.get('django_scaler_percentage_slowest')


def redirect_regexes_dummy():
    return []


def redirect_regexes_from_cache():
    return cache.get('django_scaler_regexes')


class ScalerMiddleware:
    """Add as the first middleware in your settings file"""

    def process_request(self, request):

        # Ajax requests are not subject to scaling. Busy page is exempt from
        # scaling.
        if request.is_ajax() or request.META['PATH_INFO'] == SERVER_BUSY_URL:
            return

        # If a n_slowest or percentage_slowest is provided then forcefully
        # redirect the n slowest or percentage_slowest requests. This allows
        # external processes to easily instruct us to scale back.
        n_slowest = settings.DJANGO_SCALER.get(
            'redirect_n_slowest_function', redirect_n_slowest_dummy
        )()
        percentage_slowest = settings.DJANGO_SCALER.get(
            'redirect_percentage_slowest_function',
            redirect_percentage_slowest_dummy
        )()
        regexes = settings.DJANGO_SCALER.get(
            'redirect_regexes_function',
            redirect_regexes_dummy
        )()
        if not request.is_ajax():
            if n_slowest or percentage_slowest:
                # Sort by slowest reversed
                paths = sorted(
                    _request_response_times,
                    key=_request_response_times.__getitem__,
                    reverse=True
                )
            if n_slowest:
                li = paths[:n_slowest]
                if request.META['PATH_INFO'] in li:
                    return HttpResponseRedirect(SERVER_BUSY_URL)
            if percentage_slowest:
                n = int(round(percentage_slowest / 100.0 * len(paths)))
                li = paths[:n]
                if request.META['PATH_INFO'] in li:
                    return HttpResponseRedirect(SERVER_BUSY_URL)
            if regexes:
                for regex in regexes:
                    m = re.match(r'%s' % regex, request.META['PATH_INFO'])
                    if m is not None:
                        return HttpResponseRedirect(SERVER_BUSY_URL)

        # On to automatic redirection
        now = time.time()

        # Marker for process_response
        setattr(request, '_django_scaler_stamp', now)

        # Cache key uses path info
        prefix = request.META['PATH_INFO'] + '-scaler-'

        # Fetch values
        key_stamp = prefix + 'stamp'
        key_hits = prefix + 'hits'
        key_trend = prefix + 'trend'
        key_redir = prefix + 'redir'
        stamp = _cache.get(key_stamp, 0)
        hits = _cache.get(key_hits, 0)
        trend = _cache.get(key_trend, [])
        redir = _cache.get(key_redir, now)

        # Nothing to do if not enough hits yet
        if hits > settings.DJANGO_SCALER.get('trend_size', 100):
            avg = stamp * 1.0 / hits

            # Update request response times dictionary
            _request_response_times[request.META['PATH_INFO']] = avg

            # If trend is X slower than average then redirect, unless
            # enough time has passed to attempt processing.
            slow_threshold = settings.DJANGO_SCALER.get(
                'slow_threshold', 4.0
            )
            if sum(trend) * 1.0 / len(trend) > avg * slow_threshold:

                # Has enough time passed to allow the request?
                redirect_for = settings.DJANGO_SCALER.get(
                    'redirect_for', 60
                )
                if now - redir > redirect_for:
                    # Yes, enough time has passed

                    # Clear time of last redirect
                    try:
                        del _cache[key_redir]
                    except KeyError:
                        pass

                    # Clear trend since it currently stores slow response
                    # times. We want a fresh start.
                    _cache[key_trend] = []

                else:
                    # No, not enough time has passed. Keep redirecting.

                    # Remove marker so process_response does not store data
                    delattr(request, '_django_scaler_stamp')

                    # Set time of last redirect if it has not been set
                    _cache.setdefault(key_redir, now)

                    return HttpResponseRedirect(SERVER_BUSY_URL)

    def process_response(self, request, response):
        t = getattr(request, '_django_scaler_stamp', None)
        # Anything to do?
        if t is not None:
            # Diff in milliseconds
            diff = int((time.time() - t) * 1000)

            # Fetch values
            prefix = request.META['PATH_INFO'] + '-scaler-'
            key_stamp = prefix + 'stamp'
            key_hits = prefix + 'hits'
            key_trend = prefix + 'trend'
            stamp = _cache.get(key_stamp, 0)
            hits = _cache.get(key_hits, 0)
            trend = _cache.get(key_trend, [])

            # Set values
            _cache[key_stamp] = stamp + diff
            _cache[key_hits] = hits + 1
            trend_size = settings.DJANGO_SCALER.get('trend_size', 100)
            _cache[key_trend] = (trend + [diff])[-trend_size:]

        return response
